[Spark] Test type widening compatibility with other Delta features

johanl-db opened the pull request:
## Description
Additional tests covering type widening and:
- Reading CDF
- Column mapping
- Time travel
- RESTORE
- CLONE

## How was this patch tested?
Test only

tomvanbussel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala:
```diff
@@ -505,6 +506,84 @@ trait DeltaTypeWideningNestedFieldsTests {
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN b TYPE int")
+    assert(readDeltaTable(tempPath).schema("b").dataType === IntegerType)
+    append(Seq(4).toDF("b"))
+    checkAnswer(sql(s"SELECT b FROM delta.`$tempPath`"), Seq(Row(1), Row(2), Row(3), Row(4)))
```
Do we also have a test for dropping the feature with a column that was renamed after its type was changed?

tomvanbussel commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala:
```diff
@@ -1252,6 +1399,50 @@ trait DeltaTypeWideningTableFeatureTests {
+    addSingleFile(Seq(Int.MinValue), IntegerType)
+
+    // RESTORE to version 2 -> ALTER TABLE CHANGE COLUMN TYPE.
+    // The type change is then still present and the first file initially added at version 0 is
```
What will happen when restoring to before the type change? We should still require history truncation in this case.

johanl-db replied:
I added a test to cover this. Dropping the feature will indeed fail as historical versions still contain the feature, asking to wait for the retention period to expire

tomvanbussel reviewed (commented):
Added a few suggestions for additional tests. Do we already have tests for constraints and generated columns. AFAIK we disallow type changes on generated columns, but constraints might be interesting as it's possible to create a constraint after changing the type.

johanl-db commented on spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala:
```diff
@@ -505,6 +507,84 @@ trait DeltaTypeWideningNestedFieldsTests {
+        "config" -> ".*defaultSchemaModeForColumnMappingTable"
+      )
+    )
+    checkAnswer(readCDF(start = 3, end = 3), Seq(Row(3, "insert"), Row(4, "insert")))
```
You can directly provide a snapshot to use to get the read schema, I added a test for this.

johanl-db commented:
> Added a few suggestions for additional tests. Do we already have tests for constraints and generated columns. AFAIK we disallow type changes on generated columns, but constraints might be interesting as it's possible to create a constraint after changing the type.

There are tests already for generated columns and constraints, added as part of https://github.com/delta-io/delta/pull/2881

I added a test here to cover constraints + RESTORE which is an interesting case:
1. Change column `a` type from `byte` to `int`
2. Add a CHECK constraint on column `a`.
3. RESTORE table to before the type change.

The constraint was added on type `int` and the column now has type `byte`. This does work though because CHECK constraints are part of the table metadata and it gets removed during RESTORE.

tomvanbussel reviewed (approved):
LGTM!
